nucleotide_transformer/sCellTransformer/model.py

```diff
@@ -0,0 +1,791 @@
+                    outs[dkey] = output["attention_weights"][:, map_number + 1]
+
+        return x, outs
+
+    @hk.transparent
+    def _attention_block(self, layer_idx: int) -> SelfAttentionBlock:
+        return SelfAttentionBlock(  # type: ignore
+            num_heads=self._config.attention_heads,
+            embed_dim=self._config.embed_dim,
+            key_size=self._config.key_size,
+            ffn_embed_dim=self._config.ffn_embed_dim,
+            add_bias_kv=False,
+            add_bias_fnn=False,
+            ffn_activation_name="swish",
+            use_glu_in_ffn=True,
+            rotary_embedding_config=self._rotary_embedding_config,
+            layer_norm_eps=self._config.layer_norm_eps,
+            pre_layer_norm=True,
+            name=f"attention_layer_{layer_idx}",
+        )
+
+    @hk.transparent
+    def _spatial_embedding_layer(self, tokens: tuple[Tokens]) -> Embedding:
+
+        tokens, coordinates = tokens  # type: ignore
+        batch, seq_len = tokens.shape  # type: ignore
+        x = self._embed_layer(tokens)  # type: ignore
+        coordinates_embed_layer = SpatialEncoding(
+            embed_dim=self._config.token_embed_dim,
+            num_scales=self._config.num_scales,
+            sigma_min=self._config.sigma_min,
+            sigma_max=self._config.sigma_max,
+        )
+        x_coordinates = coordinates_embed_layer(coordinates)  # type: ignore
+        x = rearrange(
+            x,
+            "batch (cells genes) dim -> batch cells genes dim",
+            cells=self._config.num_cells,
+            genes=seq_len // self._config.num_cells,
+        )
+        x = x + x_coordinates[:, :, jnp.newaxis, :]
+        x = rearrange(x, "batch cells genes dim -> batch (cells genes) dim")
+        return x
+
+    def __call__(self, tokens: Tokens | tuple[Tokens]) -> Dict[str, Embedding]:
+
+        # Prepare outputs dict
+        outs: Dict[str, Embedding] = {}
+
+        # Compute embeddings
+
+        if self._config.use_spatial_information:
+            x = self._spatial_embedding_layer(tokens)
+        else:
+            x = self._embed_layer(tokens)
+
+        # Main body
+
+        x = jnp.transpose(x, (0, 2, 1))
+        x = self.stem(x)
+        x, residuals = self.conv_tower(x)
+        x = jnp.transpose(x, (0, 2, 1))
+        outs["residuals"] = residuals
+        outs["conv_out"] = x
+
+        x, outs = self.transformer_tower(x, outs=outs, attention_mask=None)
+        outs["transformer_out"] = x
+
+        x = jnp.transpose(x, (0, 2, 1))
+        x = self.deconv_tower(x, residuals)
+        outs["deconv_out"] = x
+        x = jnp.transpose(x, (0, 2, 1))
+        outs["embedding"] = x.reshape(
+            (x.shape[0], self._config.num_cells, -1, x.shape[-1])
+        )
+
+        x = self.lm_head(x)
+        outs["logits"] = x
+
+        return outs  # type: ignore
+
+
+def build_long_range_nt_fn(
+    config: LongRangeNTConfig,
+    compute_dtype: jnp.dtype = jnp.float32,
```

Comment:
We could remove these 3 params as we haven't validated that the model supports mixed precision (in the sense that it maintains similar performance).